% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/allelecount.R
\name{dumpCounts.ICGCconsensusIndel}
\alias{dumpCounts.ICGCconsensusIndel}
\title{Dump allele counts from vcf - ICGC pancancer consensus indel pipeline}
\usage{
dumpCounts.ICGCconsensusIndel(
  vcf_infile,
  tumour_outfile,
  refence_genome = "hg19",
  dummy_alt_allele = NA,
  dummy_ref_allele = NA
)
}
\arguments{
\item{vcf_infile}{The vcf file to read in}

\item{tumour_outfile}{File to save the tumour counts to}

\item{refence_genome}{Optional parameter specifying the reference genome build used}

\item{dummy_alt_allele}{Specify allele to be used to encode the alt counts (the indel can be multiple alleles)}

\item{dummy_ref_allele}{Specify allele to be used to encode the ref counts (the indel can be multiple alleles)}
}
\description{
Dump allele counts stored in the info column of the VCF file. Output will go into a file
supplied as tumour_outfile. It will be a fully formatted allele counts file as returned 
by alleleCounter. There are no counts for the matched normal.
}
\author{
sd11
}
